processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -310,7 +331,10 @@ public void advance()
                       if (Thread.interrupted()) {
                         throw new QueryInterruptedException();
                       }
-                      cursorOffset.increment();
+
+                      do {
+                        cursorOffset.increment();
+                      } while (!isDone() && !filterMatcher.matches());
```

Comment:
this will also result in isDone being called twice for every iteration, since query engines always have to check isDone before calling advance.